#164139768 Send verification email 164139768

devskope opened the pull request:
**What does this PR do?**
Sends verification emails to users after successful registration

**Description of Task to be completed?**
- Send mail to signed up user
-  add endpoint to verify User accounts

**How should this be manually tested?**
- Checkout `ft-send-verification-email-164139768`
- Run `npm i`
- Create a new user account by sending a valid post request to `/api/users`
  A successfully created user should receive a verification email.

**Any background context you want to provide?**
N/A

**What are the relevant pivotal tracker stories?**
#164139768

**Screenshots (if appropriate)**

Commit:
feat(account verification): send account verification emails

Commit:
feat(account verification): send account verification emails

Commit:
updated swagger doc

Commit:
feat(account verification): build dynamic verification urls

- allow for dynamic hostnames

LordUche commented on server/models/User.js:
```diff
@@ -11,35 +11,36 @@ module.exports = (sequelize, DataTypes) => {
   );
   User.beforeValidate(user => {
-    user.hash = hashPassword(user.hash);
+    if (user.hash) user.hash = hashPassword(user.hash);
```
Why did you add an `if` statement here?

andela-pessien replied:
I think the aim is to avoid rehashing an already hashed password. I would advise to confirm that `user.hash` only exists here when the field is created/updated.

LordUche commented on server/utils/mailer.js:
```diff
@@ -0,0 +1,38 @@
+    from: ADMIN_EMAIL,
+    text: `Please follow this link to activate your ${siteName} account:   ${verificationLink}`,
+    html: ' ',
+    templateId: 'd-6e841534b95240fa929b2fcd78a8ff7f',
```
Where did you get this `templateId`? Is it a random string?

andela-pessien replied:
I'm assuming from context that this is a Sendgrid template - that should be made clear (no magic values please 🙂 ). Preferably would include a link to the template as well.

I like the initiative of using a template hosted online though - that allows you to make style and other changes to the email without having to redeploy your app.

devskope replied:
@andela-pessien Noted.  We'll share links in internal comms.
@LordUche  Not random string, SendGrid template ID

LordUche commented on test/users.spec.js:
```diff
@@ -117,4 +121,29 @@ describe('Authentication', () => {
+        .get(`${baseUrl}?token=io.fjf.tir`)
+        .end((err, res) => {
+          expect(res).to.have.status(500);
+          expect(res.body.verified).to.equal(undefined);
```
Why `undefined` and not `false`?

andela-pessien replied:
Seconding this

LordUche replied:
Is the user not expecting an error message?

devskope replied:
@blackdevelopa  ☝️ 
@LordUche  It is `undefined` because the `verified` property gets tacked on to the response body only when verification succeeds.
I also agree that we need to approach testing more verbosely......  if it's being sent out in the response, test it.

andela-pessien replied:
Why would the `verified` property ever be `undefined`? If verification fails then the answer to whether or not the link was `verified` is `false`, no?

devskope replied:
The latest commit updates the test and addresses this

andela-pessien commented on package.json:
```diff
@@ -52,16 +53,16 @@
     "eslint": "5.15.0",
     "eslint-config-airbnb-base": "13.1.0",
-    "eslint-config-prettier": "^4.1.0",
+    "eslint-config-prettier": "4.1.0",
```
When did we decide to use prettier? 🙂

devskope replied:
😶 @LordUche 🏃

andela-pessien commented on server/config/env-config.js:
```diff
@@ -2,11 +2,32 @@ import dotenv from 'dotenv';
   NODE_ENV,
   PORT,
-  SECRET
+  SECRET,
```
This file is getting a bit repetitive, no?

devskope replied:
This will be addressed as soon as possible.
We'll probably do a conditional export of vars based on current `NODE_ENV`

andela-pessien replied:
Please do so - I'm pretty sure the team can come up with a more efficient way than writing out the same list of variables twice. Not required for this PR, but I don't expect to still see this by the end of the week.

Also the existence of one environment variable shouldn't influence the existence of another.

andela-pessien commented on server/controllers/auth-controller.js:
```diff
@@ -33,13 +45,31 @@ class UsersController {
           });
           const { id, username } = user;
-          const token = generateToken({ id, username });
+          const [token, verificationToken] = await authTokens({
```
What's the purpose of the two tokens here?

devskope replied:
one has a shorter lifespan than the other..... they both are also configurable via external config vars

andela-pessien replied:
I am quite aware of the difference between them. I'm asking what the purpose of the (authentication) token is here.

devskope replied:
The authentication token would be required to validate user access to protected routes.

andela-pessien commented on server/utils/helpers.js:
```diff
@@ -15,6 +20,24 @@ const { SECRET } = env;
+ * @param {string} options.verificationTokenExpiresIn - verification token expiry
+ * @returns {Promise.<[string,string]>} [authToken, verificationToken]
+ */
+export const authTokens = async ({
```
Why a single function that returns both?

devskope replied:
Convenience,  seeing as both are needed in the same controller.

andela-pessien replied:
I'm not really sure how it is a convenience to generate _both_ with a single function call that requires arguments to be passed for both even though they're not both being used because auth and verification tokens are used in the same file, actually.

devskope replied:
The arguments passed for both are completely optional.  they(arguments) only exist to allow for dynamic configuration of token lifespan without app redeployment.

andela-pessien replied:
So if someone wants to create a verification token with a custom lifespan, they need to call e.g `authTokens(payload, undefined, '1h')`, which returns an array which they then destructure to take the second element as the token they want and discard the first element as they didn't actually want it in the first place but it got generated anyway.

I'm still not really sure how this is for convenience. What exactly is the use case for generating an authentication token and a verification token at the same time?

devskope replied:
This function has been dropped and the generation of the tokens has been decoupled in the latest commit

andela-pessien commented on test/users.spec.js:
```diff
@@ -117,4 +121,29 @@ describe('Authentication', () => {
+        });
+    });
+
+    it('should check that the token is valid', done => {
```
Would also like to see a test for when there's no token passed

andela-pessien replied:
Has this been implemented?

devskope replied:
Yes it has been implemented

Commit:
feat(account verification): implement account verifications

- send account verification link to user email
- verify user accounts with valid verification tokens

[Finishes #164139768]
